Verify that cert manager is installed

ANeumann82 opened the pull request:
**What this PR does / why we need it**:
- Verify that cert manager is installed before installing with web hook
- Add code to allow pre-checks before starting any installation

Fixes #1130

ANeumann82 commented:
Draft because still missing some more tests

ANeumann82 commented:
@kensipe Could you please review again?

gerred commented on pkg/kudoctl/kudoinit/types.go:
```diff
@@ -13,7 +13,16 @@ const (
 
 type InitStep interface {
+	// Returns a one-line description to print
+	Description() string
```
Consider embedding the Stringer interface instead: https://godoc.org/fmt#Stringer - this would allow this to work via fmt packages rather than having to create all of that code yourself!

ANeumann82 replied:
Done

alenkacz commented on pkg/kudoctl/kudoinit/prereq/webhook.go:
```diff
@@ -70,6 +82,46 @@ func (k kudoWebHook) AsRuntimeObjs() []runtime.Object {
+	if !strings.HasSuffix(deployment.Spec.Template.Spec.Containers[0].Image, "cert-manager-controller:v0.12.0") {
+		return kudoinit.NewWarning(fmt.Sprintf("cert-manager deployment had unexpected version. expected v0.12.0 in controller image name but found %s", deployment.Spec.Template.Spec.Containers[0].Image))
+	}
+	if deployment.Status.ReadyReplicas < 1 {
```
what do you think about using health module for this? I think it's something like `health.IsHealthy` and you can pass in deployment

ANeumann82 replied:
Good idea, done.

kensipe commented on pkg/kudoctl/kudoinit/crd/crds.go:
```diff
@@ -36,6 +36,10 @@ func NewInitializer() Initializer {
 	}
 }
 
+func (c Initializer) Description() string {
```
totally agree this should be Stringer interface... 

I'm not exactly sure how we got to the "Initializer` struct in the first place.    The `InitStep` also feels completely off.   An interface would be something you would use more than once and would be reduced in size.   If this was a `Step` I might be able to see it... but an `InitStep`.  How many and how many kinds of InitSteps are expected?

kensipe commented on pkg/kudoctl/kudoinit/result.go:
```diff
@@ -0,0 +1,98 @@
+}
+
+// NewResult initializes the Result struct
+func NewResult() Result {
```
I'm completely against this style of error handling for different reasons depending on direction.
Issues with using this direction:
1. When do we use it compared to standard go error handling?  Guidance would need to be provided.    
2. It lives in the `kudoinit` package and is NOT kudoinit specific
3. It is a copy and paste of `verifier.result.go` with no added value.   We would want to collapse this into a shared lib

I'm strongly against this altogether...   

There is a need in the verification space to not "error out" on first error but to have a full report of all errors and warnings.

For init this is standard go code to provide a function...  if we want a run time "verification" than lets build that which would have a list of all conditions reported (and perhaps called within init) but lets not have it here like this...  It doesn't make sense.  This is NOT standard go error handling.   Unless we want to take this on for all code (which I don't).

TL; DR: We need to remove this collection of errors and warnings in init or switch to using a common set of similar code to verification

ANeumann82 replied:
Moved Results to a common struct

kensipe commented on pkg/kudoctl/kudoinit/types.go:
```diff
@@ -13,7 +13,16 @@ const (
+
+	// Returns the installed artifacts as yaml manifests
 	AsYamlManifests() ([]string, error)
 	AsArray() []runtime.Object
```
It is also very confusing to reason about `kudoinit.InitStep`   It stutters on `init`.

ANeumann82 replied:
Renamed.

kensipe commented:
I would encourage adding in the needed "Verify that cert manager is installed before installing with web hook" without a complete debatable restructuring of all things init.

ANeumann82 commented:
Are you generally opposed to refactoring the init code, or just in the context of adding this feature?

Some thoughts on the structure (which I hope is generally not a bad thing, it is a bit much though, I agree on that)
- There are multiple init steps:
  - CRDs
  - PreReqNamespace
  - PreReqServiceAccount & RoleBinding
  - PreReqWebHook
  - Manager

Each of these Steps should be able to:
 - Verify that it is correctly installed ( or is outdated, looking at the KUDO upgrade)
 - Check that it can be installed (i.e. webhook needs cert-manager)
    - Some of the checks here are only warnings, as we can't fully validate the prerequisite, for example the cert-manager may run in a different namespace.
 - Install (or upgrade from the existing state)
 - Output/Return the installed resources as YAML

The Initializer should actually be the only interface there. The InitStep inside the prereqs is kind of duplicated and does the same thing, and should be flattened into the list from above.

With regards to the Result: I totally agree that copying the code is a bad thing(TM), although I have no idea to make it reusable without generics. But i'm very open to have a different pattern here.

Stringer interface makes a lot of sense, didn't know about that one.

ANeumann82 commented:
Reworked this quite a bit:
- I decided to stay with the Result return type, as I do want to verify the installation, and want warnings for the preInstall verification as well
- The actual Installation returns simple golang errors
- Moved Result into pkg/kudoctl/verify and removed the duplicate code
- Renamed InitStep to Step
- Split up the Step interface into multiple smaller ones and used composition. Was a bit unsure about this, but it does make the different responsibilities more clear

There is one more refactoring that I would like to make, in the kudoinit/prereqs package, but I'll leave that for another PR.

ANeumann82 commented:
I've just tried to rebasing this to fix the DCO things in the early commits. It did... not go well. So for this PR i'd probably manually override the DCO check

kensipe commented:
the simple solution to the DCO issue is to squash the commits... they get squash anyway on merge (by our convention), so if we are ready to merge it makes sense.
I rebased to latest master with a squash on all commits.   Now to review.

ANeumann82 commented:
Thanks for the squash and the comment. I do get the issue you have with that, and I hoped that the current state does take care of it:
- The `verify.Result` is only used in the InstallVerify interface: This will contain methods that will *only* do verification and will not do any modifications.
- For the actual `Init` action, there is a simple `error` result, as I do agree that the `verify.Result` shouldn't be used for the standard error flow.

There is just the thing that the init process should have - in my opionion - a verification step, that provides a list of warnings and errors before doing any potentially irreversable actions on the cluster. And I feel for this verification step the `verify.Result` is a good fit.

Can we agree on something like "`verify.Result` should only be used in non-modifying methods"?

If you have any constructive ideas how to handle this, please let me know. I can write some ideas that I have:

- Get rid of the `verify.Result` in the `PreInstallVerify` and return an error that contains either warnings or errors wrapped in a golang `error type. Might be an option, but a custom error would get very close to the existing `verify.Result`
- Don't return warnings, only return the first error: Could work too, but limits us severly, in my opinion. 
- Separate the verification and the actual installation/upgrade more (than just a different method/interface): Would be quite hard, as a lot of the same code is used by both, but might be possible.
- ...
